// lib0/dist/indexeddb.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require('./promise-afe9dfc3.cjs');
require('./error-8582d695.cjs');
var indexeddb = require('./indexeddb-8d495848.cjs');
require('./time-bc2081b9.cjs');
require('./metric.cjs');
require('./math-08e068f9.cjs');



exports.add = indexeddb.add;
exports.addAutoKey = indexeddb.addAutoKey;
exports.count = indexeddb.count;
exports.createIDBKeyRangeBound = indexeddb.createIDBKeyRangeBound;
exports.createIDBKeyRangeLowerBound = indexeddb.createIDBKeyRangeLowerBound;
exports.createIDBKeyRangeUpperBound = indexeddb.createIDBKeyRangeUpperBound;
exports.createStores = indexeddb.createStores;
exports.del = indexeddb.del;
exports.deleteDB = indexeddb.deleteDB;
exports.get = indexeddb.get;
exports.getAll = indexeddb.getAll;
exports.getAllKeys = indexeddb.getAllKeys;
exports.getAllKeysValues = indexeddb.getAllKeysValues;
exports.getFirstKey = indexeddb.getFirstKey;
exports.getLastKey = indexeddb.getLastKey;
exports.getStore = indexeddb.getStore;
exports.iterate = indexeddb.iterate;
exports.iterateKeys = indexeddb.iterateKeys;
exports.openDB = indexeddb.openDB;
exports.put = indexeddb.put;
exports.queryFirst = indexeddb.queryFirst;
exports.rtop = indexeddb.rtop;
exports.transact = indexeddb.transact;
